### PR TITLE
Laboratory work #4, Galkina Ksenia - 19FPL2

### PR DESCRIPTION
**Pull request by galkinaksenia**

### CODE REVIEW

File: lab_4/main.py [hunk 1]
```diff
@@ -1,74 +1,194 @@
...
+        if not isinstance(context, tuple) or not context:
+            raise ValueError
+        generated_sentence = []
+        if self._word_storage.get_id('<END>') not in context:
+            for word in context[:self._n_gram_trie.size-1]:
+                generated_sentence.append(word)
+        while self._word_storage.get_id('<END>') not in generated_sentence:
+            generated_word = self._generate_next_word(context)
+            context = context[:self._n_gram_trie.size-1] + (generated_word,)
+            context = context[-(self._n_gram_trie.size - 1):]
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: is there a way to join this and the previous lines?

File: lab_4/main.py [hunk 2]
```diff
@@ -1,74 +1,194 @@
...
+            generated_sentence.append(generated_word)
+            if len(generated_sentence) == 20:
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: read the description once more, it should be a bit different

File: lab_4/main.py [hunk 3]
```diff
@@ -1,74 +1,194 @@
...
+                generated_sentence.append(self._word_storage.get_id('<END>'))
+                return tuple(generated_sentence)
+        return tuple(generated_sentence)
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: same lines? 97-98

File: lab_4/main.py [hunk 4]
```diff
@@ -1,74 +1,194 @@
...
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if not isinstance(context, tuple) or not context:
+            raise ValueError
+        for word in context:
+            if word not in self._word_storage.storage.values():
+                raise ValueError
+        trie_size_max = []
+        for trie in self._n_gram_tries:
```
Comment (ValeriyaKuznetsova) [inline lab_4/main.py]: are we going from max size to min size here?

File: lab_4/start.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+Text generator
+"""
+
+import unittest
+from lab_4.ngrams.ngram_trie import NGramTrie
+from lab_4.main import WordStorage, LikelihoodBasedTextGenerator, BackOffGenerator
+from lab_4.main import encode_text, decode_text
+
+
+class TestsForGenerators(unittest.TestCase):
```
Comment (ValeriyaKuznetsova) [inline lab_4/start.py]: put tests in test files

Comment (ValeriyaKuznetsova) [inline lab_4/start.py]: GeneratorTest

File: lab_4/start.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        context = (storage.get_id('his'),
+                   storage.get_id('name'),
+                   storage.get_id('bruno'),)
+
+        first_generated = storage.get_id('is')
+        last_generated = storage.get_id('<END>')
+
+        generator = BackOffGenerator(storage, trie, four)
+        actual = generator._generate_sentence(context)
+        self.assertEqual(actual[2], first_generated)
```
Comment (ValeriyaKuznetsova) [inline lab_4/start.py]: cheating, be creative, create different tests (the change in indexes doesn't count)

File: lab_4/start.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+        context = (storage.get_id('his'),
+                   storage.get_id('name'),
+                   storage.get_id('is'),
+                   storage.get_id('bruno'),)
+
+        generator = BackOffGenerator(storage, five, trie, four)
+
+        text = generator.generate_text(context, 3)
+        actual = decode_text(storage, text)
```
Comment (ValeriyaKuznetsova) [inline lab_4/start.py]: same

**Review by ValeriyaKuznetsova — COMMENTED** [review]
correct these issues
you are almost there :)

File: lab_4/ngram_text_generator_score_six_test.py [hunk 8]
```diff
@@ -200,3 +200,83 @@ def test_text_generator_throws_errors(self):
...
+                   storage.get_id('<END>'))
+
+        generator = NGramTextGenerator(storage, trie)
+        actual = generator._generate_sentence(context)
+
+        self.assertEqual(1, actual.count(storage.get_id('<END>')))
+
+    def test_text_generator_generate_long_text(self):
+        """
+        should generate a bigger text and do crash
```
Comment (ValeriyaKuznetsova) [inline lab_4/ngram_text_generator_score_six_test.py]: if it crashes, why is the number of ends not 20?

**Review by ValeriyaKuznetsova — COMMENTED** [review]
have one question about tests
everything else is OK

File: lab_4/ngram_text_generator_score_six_test.py (hunk 8 above)
Comment (galkinaksenia) [reply lab_4/ngram_text_generator_score_six_test.py]: I forgot to add "not" in the description, it should not crash))

Comment (ValeriyaKuznetsova) [reply lab_4/ngram_text_generator_score_six_test.py]: Is it an adequate big text with different words or do the same words repeat? if the second variant, then it's a wrong test

**Comment by ValeriyaKuznetsova** [comment]
still have a question about one test

File: lab_4/ngram_text_generator_score_six_test.py (hunk 8 above)
Comment (galkinaksenia) [reply lab_4/ngram_text_generator_score_six_test.py]: Yes, words repeated, so i made a new test instead

**Pull request closed without merging**